### PR TITLE
Add get_slack_matrix_from_circuit_matrix() function for dynamical decoupling

### LINKED ISSUE #1161: Implement get_slack_matrix function for dynamical decoupling.
Implement the `get_slack_matrix_from_circuit_mask()` and `get_slack_matrix()` function described in the [RFC for dynamical decoupling](https://docs.google.com/document/d/1cRwFCTn6kUjI1P0kNydtevxIYtE4r8Omd_iWK0Pe8qo/edit?usp=sharing). 

They are necessary to determine the size of slack windows (holes with idling qubits) in a circuit, in order to fill them with ddd sequences.

An implementation skeleton with pseudo-code is present in the [RFC](https://docs.google.com/document/d/1cRwFCTn6kUjI1P0kNydtevxIYtE4r8Omd_iWK0Pe8qo/edit?usp=sharing) but, of course,  a different approach can be used.

### PR DESCRIPTION
**Pull request by andreamari**
Description
-----------

Adds get_slack_matrix_from_circuit_mask for digital dynamical decoupling.
Fixes #1161.

### CODE REVIEW

File: mitiq/ddd/insertion.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    Returns: The matrix of slack lengths.
+    """
+    _validate_integer_matrix(mask)
+    if not (mask**2 == mask).all():
+        raise ValueError("The input matrix elements must be 0 or 1.")
+
+    num_raws, num_cols = mask.shape
+    slack_matrix = np.zeros((num_raws, num_cols))
+    for r in range(num_raws):
```
Comment (Aaron-Robertson) [inline mitiq/ddd/insertion.py:47]:
```suggestion
    num_rows, num_cols = mask.shape
    slack_matrix = np.zeros((num_rows, num_cols))
    for r in range(num_rows):
```

File: mitiq/ddd/insertion.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        mask: The mask matrix of a quantum circuit.
+
+    Returns: The matrix of slack lengths.
+    """
+    _validate_integer_matrix(mask)
+    if not (mask**2 == mask).all():
+        raise ValueError("The input matrix elements must be 0 or 1.")
+
+    num_raws, num_cols = mask.shape
+    slack_matrix = np.zeros((num_raws, num_cols))
```
Comment (Aaron-Robertson) [inline mitiq/ddd/insertion.py:46]: Do we want to specify the slack matrix dtype as int? i.e. `slack_matrix = np.zeros((num_raws, num_cols), int)`

**Review by Aaron-Robertson — COMMENTED** [review]
Overall, LGTM @andreamari! Couple comments, but tested in a notebook it works nicely with `get_circuit_mask()` 🚀

File: mitiq/ddd/insertion.py (hunk 2 above)
Comment (andreamari) [reply mitiq/ddd/insertion.py]: Good point! Done here: [8586900](https://github.com/unitaryfund/mitiq/pull/1177/commits/858690092e75deb721fd3bd58ea15396d5a61e36)

**Review by Aaron-Robertson — APPROVED** [review]
LGTM, nice work @andreamari!

**Pull request merged**